test/integration/backward_compatible/serial/ClientClusterReconnectionRetryTest.js

```diff
@@ -0,0 +1,102 @@
+
+/**
+ * Basic tests for reconnection to cluster scenarios.
+ */
+describe('ClientClusterReconnectionRetryTest', function () {
+    let cluster;
+    let client;
+    const ASSERTION_MILISECONDS = 30000;
+    const INT32_MAX_VALUE = 2147483647;
+
+    const testFactory = new TestUtil.TestFactory();
+
+    beforeEach(function () {
+       client = undefined;
+       cluster = undefined;
+    });
+
+    afterEach(async function () {
+        await testFactory.shutdownAll();
+    });
+
+    it('testClientShouldNotTryToConnectCluster_whenThereIsNoConnection', async function () {
+        cluster = await testFactory.createClusterForSerialTests();
+        const member = await RC.startMember(cluster.id);
+        // Sleep indefinitely after the first connection attempt fails
+        client = await testFactory.newHazelcastClientForSerialTests({
+            clusterName: cluster.id,
+            connectionStrategy: {
+                connectionRetry: {
+                    initialBackoffMillis: INT32_MAX_VALUE,
+                    maxBackoffMillis: INT32_MAX_VALUE
+                }
+            }
+        });
+        await RC.terminateMember(cluster.id, member.uuid);
+        // Wait until the client-side connection is closed
+        await TestUtil.waitForConnectionCount(client, 0);
+        // Wait a bit more to make it more likely that the first reconnection
+        // attempt is made before we restart the instance
+        await TestUtil.promiseLater(ASSERTION_MILISECONDS, () => { });
+
+        await RC.startMember(cluster.id);
+        const clientConnectionsFn = await TestUtil.getClientConnections(client);
+        await TestUtil.assertTrueAllTheTime(async () => {
+            expect(clientConnectionsFn().length).to.be.equal(0);
+        }, 100, ASSERTION_MILISECONDS);
+    });
+
+    it('testClientState_AfterDisconnected', async function () {
+        const fnGetOrConnectToMember = sandbox.replace(
+            ConnectionManager.prototype,
+            'getOrConnectToMember',
+            sandbox.fake(ConnectionManager.prototype.getOrConnectToMember)
+        );
+
+        cluster = await testFactory.createClusterForSerialTests();
+        const member = await RC.startMember(cluster.id);
+        // Sleep indefinitely after the first connection attempt fails
+        client = await testFactory.newHazelcastClientForSerialTests({
+            clusterName: cluster.id,
+            connectionStrategy: {
+                connectionRetry: {
+                    initialBackoffMillis: INT32_MAX_VALUE,
+                    maxBackoffMillis: INT32_MAX_VALUE
+                }
+            }
+        });
+        await RC.terminateMember(cluster.id, member.uuid);
+        // Wait until the client-side connection is closed
+        await TestUtil.waitForConnectionCount(client, 0);
+
+        // After terminateMember client state will be DISCONNECTED_FROM_CLUSTER
+        expect(client.getConnectionManager().getConnectionRegistry().getClientState()
```

Comment:
also note that we are asserting we don't try to reconnect for 30 seconds by checking connection count. It should be enough to make sure that we don't try to reconnect.

Reply:
> I think this expect makes sense but the getOrConnectToMember can be flaky, we should not rely on that. Also we don't need a new test. We can put the expect in the test testClientShouldNotTryToConnectCluster_whenThereIsNoConnection

First of al, I didn't want to change the logic on Java ported test. But I thought about that, this test is not enough for ensuring client do not try to reconnect again. 
When I check the code `reconnectToMembers` methods call `getOrConnectToMember` funtion. So If we check the call count `getOrConnectToMember`, it is sufficient to ensure that client will not try to reconnect after the status was changed to 'DISCONNECTED_FROM_CLUSTER'. And also `getOrConnectToMember` is not flaky for this situation because I replaced with the same function and I just check that it is called or not.

IMO the second test guaranteed that the logic is correct or not. But we can discuss about that.